File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/whispers.ts
/**
 * Character-to-character whispers. A sender must be a campaign player with
 * an owned character; the DM can read every whisper, but players only see
 * whispers where their owned character is the sender or the recipient.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import {
  authenticate,
  isActor,
  isApiResult,
  findCampaign,
  requireParticipant,
  findMemberByCharacterId,
  resolveOwnedCharacterId,
} from './shared.ts';

interface WhisperRow {
  campaign_id: string;
  sequence: number;
  whisper_id: string;
  from_character_id: string;
  to_character_id: string;
  text: string;
}

interface WhisperObject {
  whisper_id: string;
  from_character_id: string;
  to_character_id: string;
  text: string;
}

function toWhisperObject(row: WhisperRow): WhisperObject {
  return {
    whisper_id: row.whisper_id,
    from_character_id: row.from_character_id,
    to_character_id: row.to_character_id,
    text: row.text,
  };
}

export function createWhisper(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const fromCharacterId = resolveOwnedCharacterId(db, campaignId, actor.username);
  if (!fromCharacterId) {
    return { status: 403, body: { error: 'only a player with an owned character may send whispers' } };
  }

  const record = body as Record<string, unknown>;
  const whisperId = record?.whisper_id;
  if (typeof whisperId !== 'string' || whisperId.length === 0) {
    return { status: 400, body: { error: 'whisper_id must be a non-empty string' } };
  }
  const toCharacterId = record?.to_character_id;
  if (typeof toCharacterId !== 'string' || toCharacterId.length === 0) {
    return { status: 400, body: { error: 'to_character_id must be a non-empty string' } };
  }
  const text = record?.text;
  if (typeof text !== 'string' || text.length === 0) {
    return { status: 400, body: { error: 'text must be a non-empty string' } };
  }

  const recipientMember = findMemberByCharacterId(db, campaignId, toCharacterId);
  if (isApiResult(recipientMember)) {
    return { status: 400, body: { error: 'to_character_id must belong to a current campaign member' } };
  }

  const existing = db
    .prepare('SELECT whisper_id FROM play_campaign_whispers WHERE campaign_id = ? AND whisper_id = ?')
    .get(campaignId, whisperId);
  if (existing) {
    return { status: 409, body: { error: 'whisper_id already exists' } };
  }

  const sequenceRow = db
    .prepare('SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_whispers WHERE campaign_id = ?')
    .get(campaignId) as { max_sequence: number };

  db.prepare(
    'INSERT INTO play_campaign_whispers (campaign_id, sequence, whisper_id, from_character_id, to_character_id, text) VALUES (?, ?, ?, ?, ?, ?)',
  ).run(campaignId, sequenceRow.max_sequence + 1, whisperId, fromCharacterId, toCharacterId, text);

  return {
    status: 201,
    body: {
      whisper_id: whisperId,
      from_character_id: fromCharacterId,
      to_character_id: toCharacterId,
      text,
    } as unknown as JsonValue,
  };
}

export function listWhispers(authHeader: string | undefined, campaignId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const rows = db
    .prepare(
      'SELECT campaign_id, sequence, whisper_id, from_character_id, to_character_id, text FROM play_campaign_whispers WHERE campaign_id = ? ORDER BY sequence ASC',
    )
    .all(campaignId) as unknown as WhisperRow[];

  const isDm = campaign.owner === actor.username;
  const ownedCharacterId = isDm ? null : resolveOwnedCharacterId(db, campaignId, actor.username);

  const items = rows
    .filter(
      (row) =>
        isDm || (ownedCharacterId !== null && (row.from_character_id === ownedCharacterId || row.to_character_id === ownedCharacterId)),
    )
    .map(toWhisperObject);

  return { status: 200, body: { whispers: items } as unknown as JsonValue };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/whispers.rb
# Character-to-character whispers: private messages between owned
# characters in a campaign. The DM sees every whisper; players only see
# whispers where their own character is the sender or the recipient.

def play_whisper_payload(row)
  {
    whisper_id: row['whisper_id'],
    from_character_id: row['from_character_id'],
    to_character_id: row['to_character_id'],
    text: row['text']
  }
end

def next_play_whisper_sequence(campaign_id)
  db.execute(
    'SELECT COALESCE(MAX(sequence), 0) + 1 AS n FROM play_whispers WHERE campaign_id = ?',
    [campaign_id]
  ).first['n']
end

def find_play_member_by_username(campaign_id, username)
  db.execute(
    'SELECT * FROM play_campaign_members WHERE campaign_id = ? AND username = ?',
    [campaign_id, username]
  ).first
end

post '/v1/play/campaigns/:id/whispers' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')
  halt 403, { error: 'forbidden' }.to_json if is_owner

  sender = find_play_member_by_username(campaign['id'], user['username'])
  halt 403, { error: 'forbidden' }.to_json unless sender

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  whisper_id = body['whisper_id']
  to_character_id = body['to_character_id']
  text = body['text']

  halt 400, { error: 'invalid whisper_id' }.to_json unless whisper_id.is_a?(String) && !whisper_id.empty?
  halt 400, { error: 'invalid to_character_id' }.to_json unless to_character_id.is_a?(String) && !to_character_id.empty?
  halt 400, { error: 'invalid text' }.to_json unless text.is_a?(String) && !text.empty?

  recipient = db.execute(
    'SELECT 1 FROM play_campaign_members WHERE campaign_id = ? AND character_id = ?',
    [campaign['id'], to_character_id]
  ).first
  halt 400, { error: 'invalid to_character_id' }.to_json unless recipient

  existing = db.execute(
    'SELECT 1 FROM play_whispers WHERE campaign_id = ? AND whisper_id = ?',
    [campaign['id'], whisper_id]
  ).first
  halt 409, { error: 'whisper_id already exists' }.to_json if existing

  from_character_id = sender['character_id']
  sequence = next_play_whisper_sequence(campaign['id'])
  db.execute(
    'INSERT INTO play_whispers (campaign_id, sequence, whisper_id, from_character_id, to_character_id, text) ' \
    'VALUES (?, ?, ?, ?, ?, ?)',
    [campaign['id'], sequence, whisper_id, from_character_id, to_character_id, text]
  )

  status 201
  { whisper_id: whisper_id, from_character_id: from_character_id, to_character_id: to_character_id, text: text }.to_json
end

get '/v1/play/campaigns/:id/whispers' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  rows = db.execute(
    'SELECT * FROM play_whispers WHERE campaign_id = ? ORDER BY sequence ASC',
    [campaign['id']]
  )

  records = rows.map { |row| play_whisper_payload(row) }
  unless is_owner
    member = find_play_member_by_username(campaign['id'], user['username'])
    character_id = member && member['character_id']
    records = records.select { |w| w[:from_character_id] == character_id || w[:to_character_id] == character_id }
  end

  { whispers: records }.to_json
end
